// otel-agent-nodejs/src/instrumentations.js
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { LoggingInstrumentation } = require('@opentelemetry/instrumentation-logging');
const { MongoDBInstrumentation } = require('@opentelemetry/instrumentation-mongodb');
const { RedisInstrumentation } = require('@opentelemetry/instrumentation-redis');
const { ExpressInstrumentation } = require('@opentelemetry/instrumentation-express');
const { KoaInstrumentation } = require('@opentelemetry/instrumentation-koa');

function loadInstrumentations() {
  return [
    getNodeAutoInstrumentations({
      '@opentelemetry/instrumentation-http': {
        enabled: true,
        ignoreIncomingPaths: ['/health', '/metrics'], // Ajuste conforme necessário
      },
    }),
    new LoggingInstrumentation({
      logHook: (span, record) => {
        span.setAttribute('log.message', record.message);
        span.setAttribute('log.level', record.level);
      },
    }),
    new MongoDBInstrumentation({
      enhancedDatabaseReporting: true, // Mais detalhes ao trace
    }),
    new RedisInstrumentation({
      dbStatementSerializer: (cmdName, cmdArgs) => `${cmdName} [${cmdArgs.join(', ')}]`,
    }),
    new ExpressInstrumentation(),
    new KoaInstrumentation(),
  ];
}

module.exports = { loadInstrumentations };
